### ISSUE TITLE
Type annotations

### ISSUE THREAD

**Issue by TilmanK** [pytest-dev/pytest-qt#417]
I stumbled upon this when setting a new project where I haven't configured mypy yet to just ignore pytest-qt: There are no type annotations. 

Is there a reason from this apart from “It's not there because no one did it yet”? It would be a nice-to-have thing, since the earliest mayor Python version supported is 3.6 and pytest itself has them (as far as I remember)…

**Comment by nicoddemus** [comment]
Hi @TilmanK! 

>  “It's not there because no one did it yet”?

Mainly that only. I'm a big fan of having type annotations (even aiming for full type annotations), but nobody has stopped to bring them to pytest-qt yet.

**Comment by TilmanK** [comment]
Hmm, any idea how to address the dynamic use of the Qt library? Maybe using TypeVars in some way?

**Comment by nicoddemus** [comment]
> the dynamic use of the Qt library?

Hmm sorry can you be more specific?

**Comment by TilmanK** [comment]
Sure. Let's take `QtBot.addWidget()` as an example:

'''py
    def addWidget(self, widget, *, before_close_func=None):
´´´

How do I annotate that? I can use `widget: qt_api.QtWidgets.QWidget`. Doing so results in an error: `Name "qt_api.QtWidgets.QWidget" is not defined`

**Comment by nicoddemus** [comment]
Oh I see thanks.

Hmm that's tricky, not sure how to overcome that: `qt_api` by design only defines the symbols at runtime (during `pytest_configure`), so there's no way to do that at static time.

**Comment by TilmanK** [comment]
We could do something like.

```py
if TYPE_CHECKING:
    qt_api = PyQt6
```

Maybe...

**Comment by nicoddemus** [comment]
I thought of that, the problem is that the choice of using PyQt or PySide is defined at realtime too, so that snippet won't work for PySide users... :/

**Comment by TilmanK** [comment]
> I thought of that, the problem is that the choice of using PyQt or PySide is defined at realtime too, so that snippet won't work for PySide users... :/

Yes, I know. Sorry, the code snippet isn't clear.  Of course, I don't want to hard-wire the API to PyQt6, I was just thinking about a first approach to tell mypy which package to use...

**Comment by nicoddemus** [comment]
> I was just thinking about a first approach to tell mypy which package to use...

**Comment by nicoddemus** [comment]
Just thinking about this: we might decide to type any qt-related class as `Any` for now, and properly type everything else. Having a `widget: Any` parameter is not so bad for now if we type the other parameters. If we find a solution down the road, we can then improve the typing coverage accordingly.

**Comment by TilmanK** [comment]
I already started to work on a PR doing exactly this 😊

**Comment by nicoddemus** [comment]
Great! 😁

**Comment by The-Compiler** [comment]
FWIW, the way other Qt wrappers seem to approach this ([qtpy](https://github.com/spyder-ide/qtpy/blob/master/qtpy/cli.py), [qts](https://github.com/python-qt-tools/qts/blob/main/src/qts/_cli.py)) is that they have a CLI supplying `--always-true` and `--always-false` [options to mypy](https://mypy.readthedocs.io/en/stable/command_line.html?highlight=always-true#cmdoption-mypy-always-true).

**Comment by adam-grant-hendry** [comment]
Thanks for linking me to the discussion here @nicoddemus ! I'll throw my hat in the ring and put up a PR.

From our discussion, we agreed adding type hints inline to the source was preferred over using stubs(`*.pyi` files), which I 100% support. I think this is the best approach.

I have one file hinted as of yet. Per [PEP 561](https://peps.python.org/pep-0561/#partial-stub-packages), you can add a line "partial\n" to the "py.typed" marker to indicate the package is partially stubbed. I'll add this so we can add type hints one-by-one to each module as we create them and have PRs for individual modules as opposed to a giant PR that type hints the whole code base at one go.

**Comment by TilmanK** [comment]
Go for it, I currently can find the time to do this anyway...

**Comment by herobank110** [comment]
Is there any status update on adding the type annotations?

**Comment by The-Compiler** [comment]
No, otherwise it would be in this issue, that's kind of the whole point of a public issue tracker 😉

**Comment by herobank110** [comment]
mind if I take a stab at it?

**Comment by herobank110** [comment]
I have a proposed proof of concept: https://github.com/pytest-dev/pytest-qt/pull/605.
The main idea is using TYPE_CHECKING to import the relevant qt code from `qtpy`, as I'm not sure the qt_compat/qt_api in this repo will be able to provide much type annotation capabilities. Technically qtpy wouldn't need to be added to this project's dependencies - users of the library can have it in their own project to get the typing. However for mypy to actually type check in this project's CI properly, perhaps qtpy could be added as a dev dependency?

Also I'm not aware of a way to properly type *args and **kwargs for forwarding, as needed in the mouseClick, etc. functions. One solution may be to duplicate the parameters from the Qt API.

I have added mypy to the precommit config and it seems to be running on CI
